File: apps/backend/form/assessment.py
from wtforms import StringField, IntegerField, FloatField, TextAreaField, BooleanField, SelectField, FieldList, FormField
from wtforms.validators import DataRequired, Optional, Length, NumberRange, ValidationError
from .base import BaseForm


class AssessmentQueryForm(BaseForm):
    """测评查询表单"""
    page = IntegerField('页码', [Optional()], default=1)
    per_page = IntegerField('每页数量', [Optional()], default=10)
    name = StringField('测评名称', [Optional()])
    category = StringField('测评分类', [Optional()])
    difficulty = SelectField('难度', [Optional()], choices=[
        ('', '全部'),
        ('easy', '简单'),
        ('medium', '中等'),
        ('hard', '困难')
    ], default='')
    is_free = BooleanField('是否免费', [Optional()])
    status = SelectField('状态', [Optional()], choices=[
        ('', '全部'),
        ('draft', '草稿'),
        ('published', '已发布'),
        ('archived', '已归档')
    ], default='')
    keyword = StringField('搜索关键词', [Optional()])
    sort_by = StringField('排序字段', [Optional()], default='create_time')
    sort_order = StringField('排序方向', [Optional()], default='desc')


class AssessmentCreateForm(BaseForm):
    """创建测评表单"""
    name = StringField('测评名称', [DataRequired(), Length(1, 200)])
    subtitle = StringField('测评副标题', [Optional(), Length(0, 300)])
    description = TextAreaField('测评描述', [Optional()])
    cover_image = StringField('封面图片', [Optional(), Length(0, 255)])
    price = FloatField('价格', [Optional(), NumberRange(min=0)], default=0.0)
    original_price = FloatField('原价', [Optional(), NumberRange(min=0)], default=0.0)
    duration = IntegerField('预计时长(分钟)', [Optional(), NumberRange(min=1, max=1440)], default=30)
    difficulty = SelectField('难度', [Optional()], choices=[
        ('easy', '简单'),
        ('medium', '中等'),
        ('hard', '困难')
    ], default='medium')
    category = StringField('测评分类', [Optional(), Length(0, 50)])
    status = SelectField('状态', [Optional()], choices=[
        ('draft', '草稿'),
        ('published', '已发布'),
        ('archived', '已归档')
    ], default='draft')
    tags = StringField('标签', [Optional()])  # 逗号分隔的字符串
    instructions = TextAreaField('测评说明', [Optional()])
    is_free = BooleanField('是否免费', [Optional()], default=True)
    sort_order = IntegerField('排序', [Optional()], default=0)


class AssessmentUpdateForm(BaseForm):
    """更新测评表单"""
    name = StringField('测评名称', [Optional(), Length(1, 200)])
    subtitle = StringField('测评副标题', [Optional(), Length(0, 300)])
    description = TextAreaField('测评描述', [Optional()])
    cover_image = StringField('封面图片', [Optional(), Length(0, 255)])
    price = FloatField('价格', [Optional(), NumberRange(min=0)])
    original_price = FloatField('原价', [Optional(), NumberRange(min=0)])
    duration = IntegerField('预计时长(分钟)', [Optional(), NumberRange(min=1, max=1440)])
    difficulty = SelectField('难度', [Optional()], choices=[
        ('easy', '简单'),
        ('medium', '中等'),
        ('hard', '困难')
    ])
    category = StringField('测评分类', [Optional(), Length(0, 50)])
    status = SelectField('状态', [Optional()], choices=[
        ('draft', '草稿'),
        ('published', '已发布'),
        ('archived', '已归档')
    ])
    tags = StringField('标签', [Optional()])  # 逗号分隔的字符串
    instructions = TextAreaField('测评说明', [Optional()])
    is_free = BooleanField('是否免费', [Optional()])
    sort_order = IntegerField('排序', [Optional()])


class AssessmentOptionForm(BaseForm):
    """测评选项表单"""
    option_text = StringField('选项内容', [DataRequired(), Length(1, 500)])
    option_value = StringField('选项值', [Optional(), Length(0, 50)])
    score = FloatField('选项分值', [Optional()], default=0.0)
    option_order = IntegerField('选项序号', [Optional()], default=0)


class AssessmentQuestionForm(BaseForm):
    """测评题目表单"""
    question_text = TextAreaField('题目内容', [DataRequired()])
    question_type = SelectField('题目类型', [DataRequired()], choices=[
        ('single', '单选'),
        ('multiple', '多选'),
        ('text', '文本'),
        ('scale', '量表')
    ], default='single')
    question_order = IntegerField('题目序号', [Optional()], default=0)
    is_required = BooleanField('是否必答', [Optional()], default=True)
    score_weight = FloatField('分值权重', [Optional(), NumberRange(min=0)], default=1.0)
    dimension = StringField('维度标识', [Optional(), Length(0, 50)])
    description = TextAreaField('题目说明', [Optional()])
    options = FieldList(FormField(AssessmentOptionForm), min_entries=0)


class AssessmentQuestionCreateForm(BaseForm):
    """创建测评题目表单"""
    assessment_id = StringField('测评ID', [DataRequired(), Length(1, 50)])
    question_text = TextAreaField('题目内容', [DataRequired()])
    question_type = SelectField('题目类型', [DataRequired()], choices=[
        ('single', '单选'),
        ('multiple', '多选'),
        ('text', '文本'),
        ('scale', '量表')
    ], default='single')
    question_order = IntegerField('题目序号', [Optional()], default=0)
    is_required = BooleanField('是否必答', [Optional()], default=True)
    score_weight = FloatField('分值权重', [Optional(), NumberRange(min=0)], default=1.0)
    dimension = StringField('维度标识', [Optional(), Length(0, 50)])
    description = TextAreaField('题目说明', [Optional()])


class AssessmentQuestionUpdateForm(BaseForm):
    """更新测评题目表单"""
    question_text = TextAreaField('题目内容', [Optional()])
    question_type = SelectField('题目类型', [Optional()], choices=[
        ('single', '单选'),
        ('multiple', '多选'),
        ('text', '文本'),
        ('scale', '量表')
    ])
    question_order = IntegerField('题目序号', [Optional()])
    is_required = BooleanField('是否必答', [Optional()])
    score_weight = FloatField('分值权重', [Optional(), NumberRange(min=0)])
    dimension = StringField('维度标识', [Optional(), Length(0, 50)])
    description = TextAreaField('题目说明', [Optional()])


class AssessmentAnswerForm(BaseForm):
    """测评答案表单"""
    question_id = StringField('题目ID', [DataRequired(), Length(1, 50)])
    option_ids = StringField('选项ID列表', [Optional()])  # JSON字符串
    answer_text = TextAreaField('文本答案', [Optional()])


class AssessmentSubmitForm(BaseForm):
    """提交测评表单"""
    assessment_id = StringField('测评ID', [DataRequired(), Length(1, 50)])
    answers = FieldList(FormField(AssessmentAnswerForm), min_entries=1)
    is_anonymous = BooleanField('是否匿名', [Optional()], default=False)


class AssessmentRecordQueryForm(BaseForm):
    """测评记录查询表单"""
    page = IntegerField('页码', [Optional()], default=1)
    per_page = IntegerField('每页数量', [Optional()], default=10)
    user_id = StringField('用户ID', [Optional()])
    assessment_id = StringField('测评ID', [Optional()])
    status = SelectField('状态', [Optional()], choices=[
        ('', '全部'),
        ('in_progress', '进行中'),
        ('completed', '已完成'),
        ('expired', '已过期')
    ], default='')
    start_date = StringField('开始日期', [Optional()])  # YYYY-MM-DD格式
    end_date = StringField('结束日期', [Optional()])  # YYYY-MM-DD格式
    sort_by = StringField('排序字段', [Optional()], default='create_time')
    sort_order = StringField('排序方向', [Optional()], default='desc')


class AssessmentStartForm(BaseForm):
    """开始测评表单"""
    assessment_id = StringField('测评ID', [DataRequired(), Length(1, 50)])
    is_anonymous = BooleanField('是否匿名', [Optional()], default=False) 

File: apps/frontend/src/api/assessment.js
/**
 * 测评相关API接口
 */
import { request } from '@/utils/request'

// 测评管理API
export const assessmentAPI = {
  // 获取测评列表
  getAssessments: (params) => {
    return request({
      url: '/assessment',
      method: 'get',
      params
    })
  },

  // 获取测评详情
  getAssessment: (assessmentId) => {
    return request({
      url: `/assessment/${assessmentId}`,
      method: 'get'
    })
  },

  // 创建测评
  createAssessment: (data) => {
    return request({
      url: '/assessment',
      method: 'post',
      data
    })
  },

  // 更新测评
  updateAssessment: (assessmentId, data) => {
    return request({
      url: `/assessment/${assessmentId}`,
      method: 'put',
      data
    })
  },

  // 删除测评
  deleteAssessment: (assessmentId) => {
    return request({
      url: `/assessment/${assessmentId}`,
      method: 'delete'
    })
  },

  // 创建测评题目
  createQuestion: (assessmentId, data) => {
    return request({
      url: `/assessment/${assessmentId}/questions`,
      method: 'post',
      data
    })
  },

  // 更新测评题目
  updateQuestion: (assessmentId, questionId, data) => {
    return request({
      url: `/assessment/${assessmentId}/questions/${questionId}`,
      method: 'put',
      data
    })
  },

  // 删除测评题目
  deleteQuestion: (assessmentId, questionId) => {
    return request({
      url: `/assessment/${assessmentId}/questions/${questionId}`,
      method: 'delete'
    })
  },

  // 获取题目选项列表
  getOptions: (assessmentId, questionId) => {
    return request({
      url: `/assessment/${assessmentId}/questions/${questionId}/options`,
      method: 'get'
    })
  },

  // 批量保存题目选项
  saveOptions: (assessmentId, questionId, options) => {
    return request({
      url: `/assessment/${assessmentId}/questions/${questionId}/options`,
      method: 'post',
      data: options
    })
  },

  // 删除选项
  deleteOption: (assessmentId, questionId, optionId) => {
    return request({
      url: `/assessment/${assessmentId}/questions/${questionId}/options/${optionId}`,
      method: 'delete'
    })
  },

  // 开始测评
  startAssessment: (data) => {
    return request({
      url: '/assessment/start',
      method: 'post',
      data
    })
  },

  // 提交测评
  submitAssessment: (data) => {
    return request({
      url: '/assessment/submit',
      method: 'post',
      data
    })
  },

  // 获取测评记录列表
  getRecords: (params) => {
    return request({
      url: '/assessment/records',
      method: 'get',
      params
    })
  },

  // 获取测评记录详情
  getRecord: (recordId) => {
    return request({
      url: `/assessment/records/${recordId}`,
      method: 'get'
    })
  }
}

// 测评查询参数构建器
export class AssessmentQueryBuilder {
  constructor() {
    this.params = {
      page: 1,
      per_page: 10
    }
  }

  // 设置分页
  page(page, perPage = 10) {
    this.params.page = page
    this.params.per_page = perPage
    return this
  }

  // 按名称搜索
  name(name) {
    if (name) {
      this.params.name = name
    }
    return this
  }

  // 按分类筛选
  category(category) {
    if (category) {
      this.params.category = category
    }
    return this
  }

  // 按难度筛选
  difficulty(difficulty) {
    if (difficulty) {
      this.params.difficulty = difficulty
    }
    return this
  }

  // 按状态筛选
  status(status) {
    if (status) {
      this.params.status = status
    }
    return this
  }

  // 是否免费
  free(isFree) {
    if (typeof isFree === 'boolean') {
      this.params.is_free = isFree
    }
    return this
  }

  // 关键词搜索
  keyword(keyword) {
    if (keyword) {
      this.params.keyword = keyword
    }
    return this
  }

  // 排序
  sort(sortBy, sortOrder = 'asc') {
    this.params.sort_by = sortBy
    this.params.sort_order = sortOrder
    return this
  }

  // 构建查询参数
  build() {
    return { ...this.params }
  }

  // 执行查询
  async execute() {
    return assessmentAPI.getAssessments(this.build())
  }
}

// 测评记录查询参数构建器
export class AssessmentRecordQueryBuilder {
  constructor() {
    this.params = {
      page: 1,
      per_page: 10
    }
  }

  // 设置分页
  page(page, perPage = 10) {
    this.params.page = page
    this.params.per_page = perPage
    return this
  }

  // 按用户ID筛选
  userId(userId) {
    if (userId) {
      this.params.user_id = userId
    }
    return this
  }

  // 按测评ID筛选
  assessmentId(assessmentId) {
    if (assessmentId) {
      this.params.assessment_id = assessmentId
    }
    return this
  }

  // 按状态筛选
  status(status) {
    if (status) {
      this.params.status = status
    }
    return this
  }

  // 日期范围
  dateRange(startDate, endDate) {
    if (startDate) {
      this.params.start_date = startDate
    }
    if (endDate) {
      this.params.end_date = endDate
    }
    return this
  }

  // 排序
  sort(sortBy, sortOrder = 'desc') {
    this.params.sort_by = sortBy
    this.params.sort_order = sortOrder
    return this
  }

  // 构建查询参数
  build() {
    return { ...this.params }
  }

  // 执行查询
  async execute() {
    return assessmentAPI.getRecords(this.build())
  }
}

// 测评表单数据构建器
export class AssessmentFormBuilder {
  constructor() {
    this.data = {
      name: '',
      subtitle: '',
      description: '',
      cover_image: '',
      price: 0,
      original_price: 0,
      duration: 30,
      difficulty: 'medium',
      category: '',
      status: 'draft',
      tags: '',
      instructions: '',
      is_free: true,
      sort_order: 0
    }
  }

  // 基本信息
  basicInfo(name, subtitle = '', description = '') {
    this.data.name = name
    this.data.subtitle = subtitle
    this.data.description = description
    return this
  }

  // 封面图片
  cover(coverImage) {
    this.data.cover_image = coverImage
    return this
  }

  // 价格信息
  pricing(price, originalPrice = null) {
    this.data.price = price
    this.data.original_price = originalPrice || price
    this.data.is_free = price === 0
    return this
  }

  // 测评设置
  settings(duration, difficulty = 'medium', category = '') {
    this.data.duration = duration
    this.data.difficulty = difficulty
    this.data.category = category
    return this
  }

  // 状态
  status(status) {
    this.data.status = status
    return this
  }

  // 标签
  tags(tags) {
    if (Array.isArray(tags)) {
      this.data.tags = tags.join(',')
    } else {
      this.data.tags = tags
    }
    return this
  }

  // 说明
  instructions(instructions) {
    this.data.instructions = instructions
    return this
  }

  // 排序
  sortOrder(order) {
    this.data.sort_order = order
    return this
  }

  // 构建表单数据
  build() {
    return { ...this.data }
  }
}

// 题目表单数据构建器
export class QuestionFormBuilder {
  constructor() {
    this.data = {
      question_text: '',
      question_type: 'single',
      question_order: 0,
      is_required: true,
      score_weight: 1.0,
      dimension: '',
      description: ''
    }
  }

  // 题目内容
  text(questionText) {
    this.data.question_text = questionText
    return this
  }

  // 题目类型
  type(questionType) {
    this.data.question_type = questionType
    return this
  }

  // 题目序号
  order(order) {
    this.data.question_order = order
    return this
  }

  // 是否必答
  required(isRequired = true) {
    this.data.is_required = isRequired
    return this
  }

  // 分值权重
  weight(weight) {
    this.data.score_weight = weight
    return this
  }

  // 维度标识
  dimension(dimension) {
    this.data.dimension = dimension
    return this
  }

  // 题目说明
  description(description) {
    this.data.description = description
    return this
  }

  // 构建表单数据
  build() {
    return { ...this.data }
  }
}

// 常量定义
export const ASSESSMENT_STATUS = {
  DRAFT: 'draft',
  PUBLISHED: 'published',
  ARCHIVED: 'archived'
}

export const ASSESSMENT_DIFFICULTY = {
  EASY: 'easy',
  MEDIUM: 'medium',
  HARD: 'hard'
}

export const QUESTION_TYPE = {
  SINGLE: 'single',
  MULTIPLE: 'multiple',
  TEXT: 'text',
  SCALE: 'scale'
}

export const RECORD_STATUS = {
  IN_PROGRESS: 'in_progress',
  COMPLETED: 'completed',
  EXPIRED: 'expired'
}

// 工具函数
export const assessmentUtils = {
  // 获取状态文本
  getStatusText(status) {
    const statusMap = {
      [ASSESSMENT_STATUS.DRAFT]: '草稿',
      [ASSESSMENT_STATUS.PUBLISHED]: '已发布',
      [ASSESSMENT_STATUS.ARCHIVED]: '已归档'
    }
    return statusMap[status] || status
  },

  // 获取难度文本
  getDifficultyText(difficulty) {
    const difficultyMap = {
      [ASSESSMENT_DIFFICULTY.EASY]: '简单',
      [ASSESSMENT_DIFFICULTY.MEDIUM]: '中等',
      [ASSESSMENT_DIFFICULTY.HARD]: '困难'
    }
    return difficultyMap[difficulty] || difficulty
  },

  // 获取题目类型文本
  getQuestionTypeText(type) {
    const typeMap = {
      [QUESTION_TYPE.SINGLE]: '单选',
      [QUESTION_TYPE.MULTIPLE]: '多选',
      [QUESTION_TYPE.TEXT]: '文本',
      [QUESTION_TYPE.SCALE]: '量表'
    }
    return typeMap[type] || type
  },

  // 获取记录状态文本
  getRecordStatusText(status) {
    const statusMap = {
      [RECORD_STATUS.IN_PROGRESS]: '进行中',
      [RECORD_STATUS.COMPLETED]: '已完成',
      [RECORD_STATUS.EXPIRED]: '已过期'
    }
    return statusMap[status] || status
  },

  // 格式化价格
  formatPrice(price) {
    if (price === 0 || price === null || price === undefined) {
      return '免费'
    }
    return `¥${price.toFixed(2)}`
  },

  // 格式化时长
  formatDuration(duration) {
    if (!duration) return '未设置'
    return `${duration}分钟`
  }
}

export default assessmentAPI
